python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> "TimestampLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.fromtimestamp(self.value / 1000000) - EPOCH).days)
+
+
+class DecimalLiteral(Literal[Decimal]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> Optional["DecimalLiteral"]:
+        if type_var.scale == abs(self.value.as_tuple().exponent):
+            return self
+        return None
+
+
+class StringLiteral(Literal[str]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(StringType)
+    def _(self, type_var: StringType) -> "StringLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.strptime(self.value, "%Y-%m-%d") - EPOCH).days)
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
+        return TimeLiteral(
+            int(
+                (
+                    datetime.datetime.strptime((EPOCH.strftime("%Y-%m-%d ") + self.value), "%Y-%m-%d %H:%M:%S.%f") - EPOCH
+                ).total_seconds()
+                * 1000000
+            )
+        )
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> TimestampLiteral:
+        try:
+            timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f")
+        except ValueError as e:
+            try:
+                timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f%z")
+                raise TypeError(f"Cannot convert String to Timestamp when timezone is included: {self.value}")
+            except:
+                raise
+
+        return TimestampLiteral(int((timestamp - EPOCH).total_seconds() * 1_000_000))
+
+    @to.register(TimestamptzType)
+    def _(self, type_var: TimestamptzType) -> "TimestampLiteral":
+        try:
+            timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f%z")
+        except ValueError as e:
+            try:
+                timestamp = datetime.datetime.strptime(self.value, "%Y-%m-%dT%H:%M:%S.%f")
+            except:
+                raise e
+            raise TypeError(f"Cannot convert String to Timestamptz, missing timezone: {self.value}")
+        utc_epoch = EPOCH.replace(tzinfo=pytz.UTC)
+        return TimestampLiteral(int((timestamp - utc_epoch).total_seconds() * 1_000_000))
+
+    @to.register(UUIDType)
+    def _(self, type_var: UUIDType) -> "UUIDLiteral":
+        return UUIDLiteral(uuid.UUID(self.value))
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> DecimalLiteral:
+        dec_val = Decimal(str(self.value))
```

Comment:
Isn't self.value already a str? I think this may hide bugs because it should be.

Reply:
yeah I removed `str` here for now, but I think this might be similar to problem above, I feel for dynamically check language like python its duck-typing nature makes very hard to assume type invariant of value